File: mobile-prototype/contract-snippets/android/ChaseMapperContractModels.kt
package org.projecthorus.chasemapper.contract

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

// MARK: - Common

@Serializable
data class ApiErrorResponse(
    val error: String,
    @SerialName("retry_after_s") val retryAfterS: Int? = null
)

// MARK: - /api/route (POST)

@Serializable
data class RouteRequest(
    @SerialName("start_lat") val startLat: Double,
    @SerialName("start_lon") val startLon: Double,
    @SerialName("end_lat") val endLat: Double,
    @SerialName("end_lon") val endLon: Double
)

@Serializable
data class RouteResponse(
    val feature: GeoJsonFeature,
    @SerialName("distance_m") val distanceM: Double,
    @SerialName("duration_s") val durationS: Double,
    val provider: String,
    @SerialName("provider_base") val providerBase: String
)

// MARK: - /api/latest_route (GET)

typealias LatestRouteResponse = GeoJsonFeature

// MARK: - /api/mobile_state (GET)

@Serializable
data class MobileStateResponse(
    @SerialName("server_time") val serverTime: String,
    val car: CarState? = null,
    val target: TargetState? = null,
    val route: RouteState,
    val eta: EtaState
)

@Serializable
data class CarState(
    val lat: Double? = null,
    val lon: Double? = null,
    val alt: Double? = null,
    val speed: Double? = null,
    val heading: Double? = null,
    @SerialName("heading_valid") val headingValid: Boolean = false,
    @SerialName("last_update") val lastUpdate: String? = null
)

@Serializable
data class TargetState(
    val callsign: String,
    val landing: LandingState,
    val telemetry: TelemetryState,
    @SerialName("time_to_landing") val timeToLanding: String? = null,
    @SerialName("time_to_landing_s") val timeToLandingS: Int? = null
)

@Serializable
data class LandingState(
    val lat: Double? = null,
    val lon: Double? = null,
    val alt: Double? = null
)

@Serializable
data class TelemetryState(
    val callsign: String? = null,
    val position: List<Double>? = null,
    @SerialName("vel_v") val velV: Double? = null,
    val speed: Double? = null,
    @SerialName("short_time") val shortTime: String? = null,
    @SerialName("packet_time") val packetTime: String? = null,
    @SerialName("time_to_landing") val timeToLanding: String? = null,
    @SerialName("server_time") val serverTime: Double? = null,
    val heading: Double? = null
)

@Serializable
data class RouteState(
    val geojson: GeoJsonFeature? = null,
    @SerialName("distance_m") val distanceM: Double? = null,
    @SerialName("duration_s") val durationS: Double? = null,
    val provider: String? = null,
    @SerialName("provider_base") val providerBase: String? = null,
    @SerialName("updated_at") val updatedAt: String? = null
)

@Serializable
data class EtaState(
    @SerialName("route_duration_s") val routeDurationS: Double? = null,
    @SerialName("payload_time_to_landing_s") val payloadTimeToLandingS: Int? = null,
    @SerialName("payload_time_to_landing") val payloadTimeToLanding: String? = null
)

// MARK: - GeoJSON (minimal contract shape)

@Serializable
data class GeoJsonFeature(
    val type: String,
    val geometry: GeoJsonGeometry,
    val properties: Map<String, JsonElement>? = null
)

@Serializable
data class GeoJsonGeometry(
    val type: String,
    val coordinates: List<List<Double>>
) {
    fun toLatLonPairs(): List<Pair<Double, Double>> =
        coordinates.mapNotNull { point ->
            if (point.size < 2) null else Pair(point[1], point[0])
        }
}

File: mobile-prototype/contract-snippets/ios/ChaseMapperContractModels.swift
import Foundation
import CoreLocation

// MARK: - Common

public struct ApiErrorResponse: Codable {
    public let error: String
    public let retryAfterS: Int?

    enum CodingKeys: String, CodingKey {
        case error
        case retryAfterS = "retry_after_s"
    }
}

public typealias LonLat = [Double]

// MARK: - /api/route (POST)

public struct RouteRequest: Codable {
    public let startLat: Double
    public let startLon: Double
    public let endLat: Double
    public let endLon: Double

    enum CodingKeys: String, CodingKey {
        case startLat = "start_lat"
        case startLon = "start_lon"
        case endLat = "end_lat"
        case endLon = "end_lon"
    }

    public init(startLat: Double, startLon: Double, endLat: Double, endLon: Double) {
        self.startLat = startLat
        self.startLon = startLon
        self.endLat = endLat
        self.endLon = endLon
    }
}

public struct RouteResponse: Codable {
    public let feature: GeoJSONFeature
    public let distanceM: Double
    public let durationS: Double
    public let provider: String
    public let providerBase: String

    enum CodingKeys: String, CodingKey {
        case feature
        case distanceM = "distance_m"
        case durationS = "duration_s"
        case provider
        case providerBase = "provider_base"
    }
}

// MARK: - /api/latest_route (GET)

public typealias LatestRouteResponse = GeoJSONFeature

// MARK: - /api/mobile_state (GET)

public struct MobileStateResponse: Codable {
    public let serverTime: String
    public let car: CarState?
    public let target: TargetState?
    public let route: RouteState
    public let eta: EtaState

    enum CodingKeys: String, CodingKey {
        case serverTime = "server_time"
        case car
        case target
        case route
        case eta
    }
}

public struct CarState: Codable {
    public let lat: Double?
    public let lon: Double?
    public let alt: Double?
    public let speed: Double?
    public let heading: Double?
    public let headingValid: Bool
    public let lastUpdate: String?

    enum CodingKeys: String, CodingKey {
        case lat
        case lon
        case alt
        case speed
        case heading
        case headingValid = "heading_valid"
        case lastUpdate = "last_update"
    }
}

public struct TargetState: Codable {
    public let callsign: String
    public let landing: LandingState
    public let telemetry: TelemetryState
    public let timeToLanding: String?
    public let timeToLandingS: Int?

    enum CodingKeys: String, CodingKey {
        case callsign
        case landing
        case telemetry
        case timeToLanding = "time_to_landing"
        case timeToLandingS = "time_to_landing_s"
    }
}

public struct LandingState: Codable {
    public let lat: Double?
    public let lon: Double?
    public let alt: Double?
}

public struct TelemetryState: Codable {
    public let callsign: String?
    public let position: [Double]?
    public let velV: Double?
    public let speed: Double?
    public let shortTime: String?
    public let packetTime: String?
    public let timeToLanding: String?
    public let serverTime: Double?
    public let heading: Double?

    enum CodingKeys: String, CodingKey {
        case callsign
        case position
        case velV = "vel_v"
        case speed
        case shortTime = "short_time"
        case packetTime = "packet_time"
        case timeToLanding = "time_to_landing"
        case serverTime = "server_time"
        case heading
    }
}

public struct RouteState: Codable {
    public let geojson: GeoJSONFeature?
    public let distanceM: Double?
    public let durationS: Double?
    public let provider: String?
    public let providerBase: String?
    public let updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case geojson
        case distanceM = "distance_m"
        case durationS = "duration_s"
        case provider
        case providerBase = "provider_base"
        case updatedAt = "updated_at"
    }
}

public struct EtaState: Codable {
    public let routeDurationS: Double?
    public let payloadTimeToLandingS: Int?
    public let payloadTimeToLanding: String?

    enum CodingKeys: String, CodingKey {
        case routeDurationS = "route_duration_s"
        case payloadTimeToLandingS = "payload_time_to_landing_s"
        case payloadTimeToLanding = "payload_time_to_landing"
    }
}

// MARK: - GeoJSON (minimal contract shape)

public struct GeoJSONFeature: Codable {
    public let type: String
    public let geometry: GeoJSONGeometry
    public let properties: [String: JSONValue]?
}

public struct GeoJSONGeometry: Codable {
    public let type: String
    public let coordinates: [LonLat]

    public var polylineCoordinates: [CLLocationCoordinate2D] {
        coordinates.compactMap { pair in
            guard pair.count >= 2 else { return nil }
            return CLLocationCoordinate2D(latitude: pair[1], longitude: pair[0])
        }
    }
}

// MARK: - Any JSON dictionary values for extensible 'properties'

public enum JSONValue: Codable {
    case string(String)
    case int(Int)
    case double(Double)
    case bool(Bool)
    case array([JSONValue])
    case object([String: JSONValue])
    case null

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let value = try? container.decode(Bool.self) {
            self = .bool(value)
        } else if let value = try? container.decode(Int.self) {
            self = .int(value)
        } else if let value = try? container.decode(Double.self) {
            self = .double(value)
        } else if let value = try? container.decode(String.self) {
            self = .string(value)
        } else if let value = try? container.decode([String: JSONValue].self) {
            self = .object(value)
        } else if let value = try? container.decode([JSONValue].self) {
            self = .array(value)
        } else {
            throw DecodingError.dataCorruptedError(in: container, debugDescription: "Unsupported JSON value")
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .null:
            try container.encodeNil()
        case .bool(let value):
            try container.encode(value)
        case .int(let value):
            try container.encode(value)
        case .double(let value):
            try container.encode(value)
        case .string(let value):
            try container.encode(value)
        case .array(let value):
            try container.encode(value)
        case .object(let value):
            try container.encode(value)
        }
    }
}
